R/PipeOpDateFeatures.R

```diff
@@ -0,0 +1,241 @@
+        dt[, (date_vars) := NULL]
+      }
+
+      if (drop_year) {
+        dt[, paste0(date_vars, ".", "year") := NULL]
+      }
+
+      dt
+    }
+  )
+)
+
+mlr_pipeops$add("datefeatures", PipeOpDateFeatures)
+
+# helper function to compute date features of a vector of dates
+compute_date_features = function(dates, features) {
+  lapply(features, FUN = function(feature) {
+    formatting = switch(feature,
+      year = "%Y",
+      month = "%m",
+      week_of_year = "%U", # starting on Sunday
+      day_of_year = "%j",
+      day_of_month = "%d",
+      day_of_week = "%w", # 0 = Sunday
+      hour = "%H",
+      minute = "%M",
+      second = "%S",
+      is_day = "%H")
+    if (feature == "is_day") {
+      hours = as.numeric(format(dates, formatting))
+      return((6L <= hours) & (hours <= 20L)) # early exit
+    }
+    as.numeric(format(dates, formatting))
+  })
+}
+
+# helper function to compute cyclic date features of date features, i.e.,
+# sine and cosine transformations of 2 * pi * x / max_x + 1
+compute_cyclic_date_features = function(date_features, features, date_var) {
+  # drop the date_var-specific colnames here, this makes it easier in lapply
+  column_names = colnames(date_features)
+  colnames(date_features) = c(column_names[1L],
+    gsub(paste0(date_var, "."), replacement = "", x = column_names[-1L]))
+  unlist(lapply(features, FUN = function(feature) {
+    # all values are expected to start at 0 and therefore may be shifted by - 1
+    value = if (feature %in% c("month", "week_of_year", "day_of_year", "day_of_month")) {
+      date_features[[feature]] - 1L
+    } else {
+      date_features[[feature]]
+    }
+    maximum = switch(feature,
+      month = 12L,
+      week_of_year = get_weeks_per_year(date_features[["year"]]),
+      day_of_year = (365L + as.integer(is_leap_year(date_features[["year"]]))),
+      day_of_month = get_days_per_month(date_features[["year"]], month = date_features[["month"]]),
+      day_of_week = 7L,
+      hour = 24L,
+      minute = 60L,
+      second = 60L)
+    value_scaled = 2L * pi * value / maximum
+    list(sin(value_scaled), cos(value_scaled))
+  }), recursive = FALSE)
+}
+
+# helper function to get the number of weeks per year, this can sometimes be 53 instead of 52 (works
+# also vectorized)
+get_weeks_per_year = function(year) {
+  as.numeric(format(as.POSIXct(paste0(year, "-12-31"), format = c("%Y-%m-%d")), "%U"))
+}
+
+# helper function to check whether the year is a leap year (works also vectorized)
+is_leap_year = function(year) {
+  (year %% 4L == 0L & year %% 100L != 0L) | (year %% 400L == 0L)
+}
+
+# helper function to get the number of days per month respecting leap years
+get_days_per_month = function(year, month) {
+  stopifnot(length(year) == length(month))
+  if (length(year) > 1L) {
+    mapply(get_days_per_month, year = year, month = month)
+  } else {
+    days_per_month = c(31L, 28L, 31L, 30L, 31L, 30L, 31L, 31L, 30L, 31L, 30L, 31L)
```

Comment:
```r
days_per_month = c(31L, 28L, 31L, 30L, 31L, 30L, 31L, 31L, 30L, 31L, 30L, 31L)
if (is_leap_year(year)) days_per_month[2] = 29
days_per_month[month]
```
does the NA handling automatically

Reply:
if year is `NA` like in the test"PipeOpDateFeatures - automatic NA handling" this does not work because is_leap_year(year) will return `NA` but I get the point - fixed in the next commit

Reply:
Hm, why can 'year' and 'month' be NA at all? Also what *should* the year == NA behaviour be? february with 28 days? 28.25? Maybe just change the value of is_leap_year(NA) to FALSE?

Reply:
`year` and `month` can be NA if the according `POSIXct` value is a complete `NA` (this can happen if a user constructs the `POSIXc` column with `optional = TRUE`) I thought that `PipeOpDateFeatures` should then also return `NA` for the datefeatures, i.e., all functions should simply pass `NA` through. As far as I know, we only have to distinguish between correctly specified `POSIXct` values (then there cannot be any `NA`'s in any fields) or complete `NA` values, then all extracted features must also be `NA`

Reply:
so in my opinion there are two possibilities to "fix" this: 1) leave as is, because it handles `NA` correctly 2) use a saver version to check whether a `POSIXct` column is `NA` and if it is, then all features are automatically set to `NA` so we must not further check whether `NA`s get passed through all functions correctly (I guess I prefer the second one)